core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -1631,26 +1637,95 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
       throw KeeperException.create(code)
   }
 
+  private def isZKSessionIdDiffFromCurrentZKSessionId(): Boolean = {
+    zooKeeperClient.sessionId != currentZooKeeperSessionId
+  }
+
+  private def isZKSessionTheEphemeralOwner(ephemeralOwnerId: Long): Boolean = {
+    ephemeralOwnerId == currentZooKeeperSessionId
+  }
+
+  private[zk] def shouldReCreateEphemeralZNode(ephemeralOwnerId: Long): Boolean = {
+    isZKSessionTheEphemeralOwner(ephemeralOwnerId) && isZKSessionIdDiffFromCurrentZKSessionId()
+  }
+
+  private def updateCurrentZKSessionId(newSessionId: Long): Unit = {
+    currentZooKeeperSessionId = newSessionId
+  }
+
   private class CheckedEphemeral(path: String, data: Array[Byte]) extends Logging {
     def create(): Code = {
       val createRequest = CreateRequest(path, data, acls(path), CreateMode.EPHEMERAL)
       val createResponse = retryRequestUntilConnected(createRequest)
+      var createResultCode = Code.OK
       createResponse.resultCode match {
-        case code@ Code.OK => code
-        case Code.NODEEXISTS => getAfterNodeExists()
+        case code@ Code.OK =>
+          createResultCode = code
+        case Code.NODEEXISTS =>
+          createResultCode = getAfterNodeExists()
         case code =>
           error(s"Error while creating ephemeral at $path with return code: $code")
+          createResultCode = code
+      }
```

Comment:
Perhaps the code can be simplified a bit as the following?

```
      val createResultCode = 
        createResponse.resultCode match {
          case code@ Code.OK =>
            code
          case Code.NODEEXISTS =>
            getAfterNodeExists()
          case code =>
            error(s"Error while creating ephemeral at $path with return code: $code")
            code
      }
```

Reply:
Yes, it definitely is simpler and allow use to use `val` instead of `var`.
Doing it.